refinery/core/models.py

```diff
@@ -1753,6 +1631,165 @@ def _create_derived_data_file_node(self, study,
+                    output_connection,
+                    output_connection.filetype,
+                    analysis_result
+                )
+                derived_data_file_node.file_uuid = (
+                    analysis_result.file_store_uuid
+                )
+                logger.debug(
+                    "Output file %s ('%s') assigned to node %s ('%s')",
+                    output_connection,
+                    analysis_result.file_store_uuid,
+                    derived_data_file_node.name,
+                    derived_data_file_node.uuid
+                )
+            output_connection.node = derived_data_file_node
+            output_connection.save()
+
+            self._link_derived_data_file_node_to_data_transformation_node(
+                graph,
+                output_connection,
+                derived_data_file_node
+            )
+
+    def _link_derived_data_file_node_to_data_transformation_node(
+            self,
+            graph,
+            output_connection,
+            derived_data_file_node
+    ):
+        """get graph edge that corresponds to this output node:
+        a. attach output node to source data transformation node
+        b. attach output node to target data transformation node
+        (if exists)"""
+        if len(graph.edges([output_connection.step])) > 0:
+            for edge in graph.edges_iter([output_connection.step]):
+                output_id = output_connection.get_output_connection_id()
+
+                if graph[edge[0]][edge[1]]['output_id'] == output_id:
+                    input_node_id = edge[0]
+                    output_node_id = edge[1]
+
+                    data_transformation_input_node = (
+                        graph.node[input_node_id]['node']
+                    )
+                    data_transformation_output_node = (
+                        graph.node[output_node_id]['node']
+                    )
+                    data_transformation_input_node.add_child(
+                        derived_data_file_node
+                    )
+                    derived_data_file_node.add_child(
+                        data_transformation_output_node
+                    )
+                    # TODO: here we could add a (Refinery internal)
+                    # attribute to the derived data file node to
+                    # indicate which output of the tool it corresponds to
+
+        # connect outputs that are not inputs for any data transformation
+        if (output_connection.is_refinery_file and
+                derived_data_file_node.parents.count() == 0):
+            graph.node[output_connection.step]['node'].add_child(
+                derived_data_file_node
+            )
+        # delete output nodes that are not refinery files and don't have
+        # any children
+        if (not output_connection.is_refinery_file and
+                derived_data_file_node.children.count() == 0):
+            output_connection.node.delete()
```

Comment:
maybe:
```python
if derived_data_file_node.children.count() == 0:
  if output_connection.is_refinery_file:
    ...
  else:
    ...
```

Reply:
I thought we could make the split here too, but looking at theses two ifs one is worried about ` derived_data_file_node.parents` and the other ` derived_data_file_node.children`